crates/hdk/src/hdk.rs

```diff
@@ -0,0 +1,180 @@
+    // Ed25519
+    fn sign(&self, sign: Sign) -> ExternResult<Signature>;
+    fn verify_signature(&self, verify_signature: VerifySignature) -> ExternResult<bool>;
+    // Entry
+    fn create(&self, entry_with_def_id: EntryWithDefId) -> ExternResult<HeaderHash>;
+    fn update(&self, update_input: UpdateInput) -> ExternResult<HeaderHash>;
+    fn delete(&self, hash: HeaderHash) -> ExternResult<HeaderHash>;
+    fn hash_entry(&self, entry: Entry) -> ExternResult<EntryHash>;
+    fn get(&self, get_input: GetInput) -> ExternResult<Option<Element>>;
+    fn get_details(&self, get_input: GetInput) -> ExternResult<Option<Details>>;
+    // Info
+    fn agent_info(&self, agent_info_input: ()) -> ExternResult<AgentInfo>;
+    fn app_info(&self, app_info_input: ()) -> ExternResult<AppInfo>;
+    fn dna_info(&self, dna_info_input: ()) -> ExternResult<DnaInfo>;
+    fn zome_info(&self, zome_info_input:()) -> ExternResult<ZomeInfo>;
+    fn call_info(&self, call_info_input: ()) -> ExternResult<CallInfo>;
+    // Link
+    fn create_link(&self, create_link_input: CreateLinkInput) -> ExternResult<HeaderHash>;
+    fn delete_link(&self, add_link_header: HeaderHash) -> ExternResult<HeaderHash>;
+    fn get_links(&self, get_links_input: GetLinksInput) -> ExternResult<Links>;
+    fn get_link_details(&self, get_links_input: GetLinksInput) -> ExternResult<LinkDetails>;
+    // P2P
+    fn call(&self, call: Call) -> ExternResult<ZomeCallResponse>;
+    fn call_remote(&self, call_remote: CallRemote) -> ExternResult<ZomeCallResponse>;
+    fn emit_signal(&self, app_signal: AppSignal) -> ExternResult<()>;
+    fn remote_signal(&self, remote_signal: RemoteSignal) -> ExternResult<()>;
+    // Random
+    fn random_bytes(&self, number_of_bytes: u32) -> ExternResult<Bytes>;
+    // Time
+    fn sys_time(&self, sys_time_input: ()) -> ExternResult<core::time::Duration>;
+    fn schedule(&self, execute_after: std::time::Duration) -> ExternResult<()>;
+    fn sleep(&self, wake_after: std::time::Duration) -> ExternResult<()>;
+    // Trace
+    fn trace(&self, trace_msg: TraceMsg) -> ExternResult<()>;
+    // XSalsa20Poly1305
+    fn create_x25519_keypair(&self, create_x25519_keypair_input: ()) -> ExternResult<X25519PubKey>;
+    fn x_salsa20_poly1305_decrypt(&self, x_salsa20_poly1305_decrypt: XSalsa20Poly1305Decrypt) -> ExternResult<Option<XSalsa20Poly1305Data>>;
+    fn x_salsa20_poly1305_encrypt(&self, x_salsa20_poly1305_encrypt: XSalsa20Poly1305Encrypt) -> ExternResult<XSalsa20Poly1305EncryptedData>;
+    fn x_25519_x_salsa20_poly1305_encrypt(&self, x_25519_x_salsa20_poly1305_encrypt: X25519XSalsa20Poly1305Encrypt) -> ExternResult<XSalsa20Poly1305EncryptedData>;
+    fn x_25519_x_salsa20_poly1305_decrypt(&self, x_25519_x_salsa20_poly1305_decrypt: X25519XSalsa20Poly1305Decrypt) -> ExternResult<Option<XSalsa20Poly1305Data>>;
+}
+
+/// The HDK implemented as externs provided by the host.
+pub struct HostHdk;
+
+impl HdkT for HostHdk {
```

Comment:
all `host_call` calls move into this trait implementation so they can be avoided with the `MockHdkT`